keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -727,20 +910,60 @@ condition makes it easier to determine if a failed pod should be restarted):
 - DeletionByTaintManager (Pod evicted by kube-controller-manager due to taints)
 - EvictionByEvictionAPI (Pod deleted by Eviction API)
 - DeletionByPodGC (an orphaned Pod deleted by pod GC)
+- DeletionByKubelet (Pod deleted due to graceful node shutdown, node pressure or Pod admission errors).
+
+We introduce a Pod condition type, called `ResourceExhausted`, used by Kubelet
+to indicate that a Pod failure is caused by exhausting pod's (or container's)
+resource (memory or ephemeral-storage). We set the `reason` field to indicate
+the reason to one of the values:
+- OOMKilled (Pod is terminated by OOM killer)
+- EphemeralStorageLimitExceeded (a limit on the Pod's emphemeral-storage is exceeded).
 
 The already existing `status.conditions` field in Pod will be used by kubernetes
-control plane components (kube-scheduler and kube-controller-manager) to append
-a dedicated condition when they send the delete operation.
+components to append a dedicated condition.
 
 The API call to append the condition will be issued as a pod status update call
 before the Pod delete request. This way the Job controller will already see the
 condition when handling a failed pod.
 
 During the implementation process we are going to review the places where the
 pod delete requests are issued to modify the code to also append a meaningful
-condition with dedicated `Type`, `Reason` and `Message` fields based on the
+condition with dedicated `type`, `reason` and `message` fields based on the
 invocation context.
 
+We list the reason constants above in order to gain the community consensus on
+informative names, but they are purely informational. It is not supported to use the `reason` or `message` fields
+when defining a pod failure policy (see `PodFailurePolicyOnPodConditionsPattern`
+in [JobSpec API](#jobspec-api)). Supporting of matching by the `reason` field
+is a possible extension of the feature which can be implemented once there is
+use case to motivate it. However, it may create a risk of breaking compatibility
+with evolving set of reasons in use, similar to the risk of
+[evolving condition types](#evolving-condition-types).
+
+### Detection and handling of exceeded limits
+
+We modify kubelet to add the `ResourceExhausted` condition if ephemeral-storage
+limits are exceeded or the Pod is terminated by Out-Of-Memory (OOM) killer.
+
+For detection of the ephemeral-storage limits being exceeded we reuse the pre-existing
+Kubelet code responsible for detection and eviction of Pods exceeding the limits (the code
+includes detection of both exceeding of the `emptyDir` size limit and pod's
+ephemeral-storage limits defined for containers). We extend the
+eviction-triggering code to add the `ResourceExhausted` condition.
```

Comment:
Can you clarify whether we will add the `ResourceExhausted` condition if the node has storage pressure but the limit is not exceeded?

We should also clarify in `Risks and Mitigations` that it's not possible to distinguish if OOMKiller is caused by limit exceeded or pressure. And the mitigation is documentation. And we recommend only using this condition with `FailJob` if requests=limits for all pods in a node. But still, even in that case, there can be cases where the pod is OOMKilled due to a system process (including kubelet itself) exhausting memory.

Reply:
Added, PTAL :)